### PR TITLE
Make deserialization generic

### PR DESCRIPTION
**Pull request by AndrejMitrovic**

### CODE REVIEW

**Comment by AndrejMitrovic** [comment]
This failure had an actual stack trace:

```
Running ./build/agora 
2620   agora                               0x000000010b9a131e _D4core7runtime18runModuleUnitTestsUZ19unittestSegvHandlerUNbiPSQCk3sys5posix6signal9siginfo_tPvZv + 62
2631   libsystem_platform.dylib            0x00007fff7d5c8f5a _sigtramp + 26
2642   agora                               0x000000010b9ae223 _D2gc4impl12conservativeQw3Gcx10smallAllocMFNbmKmkxC8TypeInfoZPv + 579
2653   agora                               0x000000010b9aa249 _D2gc4impl12conservativeQw14ConservativeGC__T9runLockedS_DQCeQCeQCcQCnQBs12mallocNoSyncMFNbmkKmxC8TypeInfoZPvS_DQEgQEgQEeQEp10mallocTimelS_DQFiQFiQFgQFr10numMallocslTmTkTmTxQCzZQFcMFNbKmKkKmKxQDsZQDl + 89
2664   agora                               0x000000010b9acf23 _DThn16_2gc4impl12conservativeQw14ConservativeGC6qallocMFNbmkxC8TypeInfoZS4core6memory8BlkInfo_ + 83
2675   agora                               0x000000010b9b2a6d gc_qalloc + 29
2686   agora                               0x000000010b9a0d12 _D4core6memory2GC6qallocFNaNbmkxC8TypeInfoZSQBqQBo8BlkInfo_ + 18
2697   agora                               0x000000010b9c44c6 _d_arrayappendcTX + 1190
2708   agora                               0x000000010b9612f4 _D3std3uni__T16UnicodeSetParserTSQBf5regex8internal6parser__T6ParserTAyaTSQCuQBpQBmQBg7CodeGenZQBiZQDi8parseSetMFNfZSQElQEk__T13InversionListTSQFlQFk8GcPolicyZQBh + 740
2719   agora                               0x000000010b960fe4 _D3std3uni7unicode__T8parseSetTSQBe5regex8internal6parser__T6ParserTAyaTSQCtQBpQBmQBg7CodeGenZQBiZQCzFNfKQCwbZSQEfQEe__T13InversionListTSQFfQFe8GcPolicyZQBh + 84
27210  agora                               0x000000010b5c64a4 _D3std5regex8internal6parser__T6ParserTAyaTSQBqQBpQBmQBg7CodeGenZQBi12parseCharsetMFNfZv + 84
27311  agora                               0x000000010b5c5ea6 _D3std5regex8internal6parser__T6ParserTAyaTSQBqQBpQBmQBg7CodeGenZQBi9parseAtomMFZv + 406
27412  agora                               0x000000010b5c5830 _D3std5regex8internal6parser__T6ParserTAyaTSQBqQBpQBmQBg7CodeGenZQBi10parseRegexMFNeZv + 1568
27513  agora                               0x000000010b5c4e3b _D3std5regex8internal6parser__T6ParserTAyaTSQBqQBpQBmQBg7CodeGenZQBi__T6__ctorTAxaZQmMFNcNeQCaQpZSQDsQDrQDoQDi__TQDeTQDaTQDaZQDq + 123
27614  agora                               0x000000010b5c4d8a _D3std5regex__T9regexImplTAyaZQpFNfQjAxaZSQBoQBn8internal2ir__T5RegexTaZQj + 106
27715  agora                               0x000000010b904439 _D3std10functional__T7memoizeS_DQBe5regex__T9regexImplTAyaZQpFNfQjAxaZSQCrQBn8internal2ir__T5RegexTaZQjVii8ZQDjFQCfQBxZQBx + 569
27816  agora                               0x000000010b5c4d07 _D3std5regex__TQjTAyaZQqFNeAQkAxaZSQBhQBg8internal2ir__T5RegexTaZQj + 551
27917  agora                               0x000000010b5c4ad4 _D3std5regex__TQjTAyaZQqFNeQjAxaZSQBgQBf8internal2ir__T5RegexTaZQj + 116
28018  agora                               0x000000010b590204 _D5dyaml8resolver18_staticCtor_L29_C1FNfZv + 1364
28119  agora                               0x000000010b9c7260 _D2rt5minfo16rt_moduleTlsCtorUZ14__foreachbody1MFKSQBx19sections_elf_shared3DSOZi + 48
28220  agora                               0x000000010b9c7c3a _D2rt19sections_elf_shared3DSO7opApplyFMDFKSQBqQBqQyZiZi + 58
28321  agora                               0x000000010b9a255c thread_entryPoint + 348
28422  libsystem_pthread.dylib             0x00007fff7d5d26c1 _pthread_body + 340
28523  libsystem_pthread.dylib             0x00007fff7d5d256d _pthread_body + 0
286Program exited with code -11
287The command "dub test --skip-registry=all --compiler=${DC}" exited with 2.
```

Although it's weird..

**Comment by AndrejMitrovic** [comment]
What if it's the Travis machine that's broken? Memory corruption?

File: source/agora/common/Deserializer.d [hunk 1]
```diff
@@ -104,30 +102,28 @@ public void deserializePart (ref long record, scope DeserializeDg dg)
...
+    ubyte[] blocks_data = serializeFull(block) ~ serializeFull(block);
 
     void deserializeArrayEntry () nothrow @safe
     {
         scope DeserializeDg dg = (size) nothrow @safe
         {
-            ubyte[] res = arrayData[0 .. size];
-            arrayData = arrayData[size .. $];
-            return res;
+            scope (exit) blocks_data = blocks_data[size .. $];
```
Comment (Geod24) [inline source/agora/common/Deserializer.d]:
```suggestion
            scope(exit) blocks_data = blocks_data[size .. $];
```

**Review by Geod24 — APPROVED** [review]
LGTM, just a tiny style comment

**Pull request merged**